source/extensions/transport_sockets/tls/ssl_handshaker.cc

```diff
@@ -95,8 +95,10 @@ Network::PostIoAction SslHandshakerImpl::doHandshake() {
     case SSL_ERROR_WANT_WRITE:
       return PostIoAction::KeepOpen;
     case SSL_ERROR_WANT_PRIVATE_KEY_OPERATION:
+      state_ = Ssl::SocketState::HandshakeInProgressByPrivateKeyOperation;
+      return PostIoAction::KeepOpen;
     case SSL_ERROR_WANT_CERTIFICATE_VERIFY:
-      state_ = Ssl::SocketState::HandshakeInProgress;
+      state_ = Ssl::SocketState::HandshakeInProgressByCertificateVerification;
```

Comment:
I wasn't sure this is the right fix after digging more.

The reason to trigger the re-entry of `ssl_do_handshake` is the async of private key operate finished and that triggers a ssl_do_handshake, then begin to verify the client certificates and the default cert validator invoke the `ssl_do_handshake` again.

The problem is the default cert validator is sync, but it still invoke the `ssl_do_handshake` to resume the handshake.

https://github.com/envoyproxy/envoy/blob/a0c50af79c7790227294f8e3a2523fd2b6255b67/source/extensions/transport_sockets/tls/ssl_handshaker.cc#L47-L56

Maybe we shouldn't determine whether this validator is async or not by `cert_validate_result_callback_.has_value()`, since it is always true as below:

https://github.com/envoyproxy/envoy/blob/a0c50af79c7790227294f8e3a2523fd2b6255b67/source/extensions/transport_sockets/tls/ssl_handshaker.cc#L61

Maybe we should have correct way to determine the validator is async or not.

Reply:
The documentation for createValidateResultCallback https://github.com/envoyproxy/envoy/blob/a0c50af79c7790227294f8e3a2523fd2b6255b67/envoy/ssl/ssl_socket_extended_info.h#L62-L66
says:
```
  /**
   * Only called when doing asynchronous cert validation.
   * @return ValidateResultCallbackPtr a callback used to return the validation result.
   */
  virtual ValidateResultCallbackPtr createValidateResultCallback() PURE;
```
So the idea is the callback is only create for async validation. How are we ending up with a callback created by sync verification?

Reply:
I think it is refactored in https://github.com/envoyproxy/envoy/pull/24297 to call this interface outside of cert validator. So the caller doesn't know if the verification will be sync or async.

Reply:
One possible fix is we know the validator is synced or not by the `doVerifyCertChain`'s return value

https://github.com/envoyproxy/envoy/blob/a0c50af79c7790227294f8e3a2523fd2b6255b67/source/extensions/transport_sockets/tls/context_impl.cc#L525-L533

Then we pass a parameter to `extended_socket_info->onCertificateValidationCompleted`, then it will know the validator is synced or not. Then decided to resume handshake or not.

Reply:
I have updated the additional description.

@RyanTheOptimist actually I think there is no sync validation any more, but the async validation may completes in the sync way.